h2/src/main/org/h2/util/ToDateTokenizer.java

```diff
@@ -0,0 +1,548 @@
+        , MONTH(PARSLET_MONTH) // Full Name of month
+        , MON(PARSLET_MONTH) // Abbreviated name of month.
+        , MM(PARSLET_MONTH) // Month (01-12; JAN = 01).
+        , RM(PARSLET_MONTH) // Roman numeral month (I-XII; JAN = I).
+        //, WW(PARSLET_Week) // Week of year (1-53)
+        //, IW(PARSLET_Week) // Week of year (1-52 or 1-53) based on the ISO standard.
+        , DDD(PARSLET_DAY) // Day of year (1-366).
+        , DAY(PARSLET_DAY) // Name of day.
+        , DD(PARSLET_DAY) // Day of month (1-31).
+        , DY(PARSLET_DAY) // Abbreviated name of day.
+        , HH24(PARSLET_TIME) //
+        , HH12(PARSLET_TIME) //
+        , HH(PARSLET_TIME) // Hour of day (1-12).
+        , MI(PARSLET_TIME) // Min
+        , SSSSS(PARSLET_TIME) // Seconds past midnight (0-86399)
+        , SS(PARSLET_TIME) //
+        , FF(PARSLET_TIME, PATTERN_FF) // Fractional seconds
+        , TZH(PARSLET_TIME) // Time zone hour.
+        , TZM(PARSLET_TIME) // Time zone minute.
+        , TZR(PARSLET_TIME) // Time zone region ID
+        , TZD(PARSLET_TIME) // Daylight savings information. Example: PST (for US/Pacific standard time);
+        , AM_PM(PARSLET_TIME, PATTERN_AM_PM) // Meridian indicator
+        , EE(PARSLET_YEAR) // NOT supported yet - Full era name (Japanese Imperial, ROC Official, and Thai Buddha calendars).
+        , E(PARSLET_YEAR) // NOT supported yet - Abbreviated era name (Japanese Imperial, ROC Official, and Thai Buddha calendars).
+        , Y(PARSLET_YEAR) //
+        , I(PARSLET_YEAR) //
+        , Q(PARSLET_MONTH) // Quarter of year (1, 2, 3, 4; JAN-MAR = 1).
+        //, W(PARSLET_Week) // Week of month (1-5)
+        , D(PARSLET_DAY) // Day of week (1-7).
+        , J(PARSLET_DAY); // NOT supported yet - Julian day; the number of days since Jan 1, 4712 BC.
+
+
+        private final static Map<Character, List<FormatTokenEnum>> cache = new HashMap<Character, List<FormatTokenEnum>>(FormatTokenEnum.values().length);
+        private final ToDateParslet toDateParslet;
+        private final Pattern patternToUse;
+
+        FormatTokenEnum(final ToDateParslet toDateParslet, final Pattern patternToUse) {
+            this.toDateParslet = toDateParslet;
+            this.patternToUse = patternToUse;
+        }
+
+        FormatTokenEnum(final ToDateParslet toDateParslet) {
+            this.toDateParslet = toDateParslet;
+            patternToUse = Pattern.compile(format("^(%s)", name()), Pattern.CASE_INSENSITIVE);
+        }
+
+        private static List<FormatTokenEnum> EMPTY_LIST = new ArrayList<FormatTokenEnum>(0);
+
+        /**
+         * OPTIMISATION: Only return a list of {@link FormatTokenEnum} that share the same 1st char
+         * using the 1st char of the 'to parse' formatStr. Or return empty list if no match.
+         */
+        static List<FormatTokenEnum> getTokensInQuestion(final ToDateParams params) {
+            List<FormatTokenEnum> result = EMPTY_LIST;
```

Comment:
this method does not need to take a full ToDateParams object. Just pass in the formatStr.